### PR TITLE
python27-bootstrap: Stop using gettext-bootstrap

### PR DESCRIPTION
**Pull request by ryandesign**
#### Description

In an effort to remove the gettext-bootstrap port, this PR changes the python27-bootstrap port to no longer use gettext-bootstrap.

See: https://trac.macports.org/ticket/58526

###### Type(s)

- [ ] bugfix
- [X] enhancement
- [ ] security fix

###### Tested on
macOS 10.13.6 17G12034
Xcode 9.4.1 9F2000

###### Verification 
Have you

- [X] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [X] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [X] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [X] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL?

- [ ] checked your Portfile with `port lint`?
- [ ] tried existing tests with `sudo port test`?
- [X] tried a full install with `sudo port -vst install`?
- [X] tested basic functionality of all binary files?

### CODE REVIEW

File: lang/python27/Portfile
Comment (akimd) [inline lang/python27/Portfile:83]: Wow...  That's the hell of a hack!

**Comment by ryandesign** [comment]
I can certainly test that. To do that, I should uninstall clang-3.4 and all its deps, and also nuke the files installed by the libcxx port, and then try to build clang-3.4 and all its deps from source, right?

**Comment by jmroot** [comment]
Right.

**Comment by kencu** [comment]
rebase please

**Comment by mascguy** [comment]
Rebased against master

**Comment by mascguy** [comment]
Bumped revisions, relative to master branch

**Comment by mascguy** [comment]
Successfully built `python27` with these changes, after first updating `gettext` to 0.21 (per sibling [PR 7399 - gettext: update to 0.21 with runtime and tools split](https://github.com/macports/macports-ports/pull/7399)).

Are we done?

**Review by mascguy — APPROVED** [review]

**Review by jmroot — CHANGES_REQUESTED** [review]
Main port shouldn't be rev bumped, and we need verification that this still works for bootstrapping (see previous review.)

**Comment by mascguy** [comment]
> Main port shouldn't be rev bumped, and we need verification that this still works for bootstrapping (see previous review.)

Per my previous comment, I built on macOS 10.6. That caused the bootstrap port to build (since 10.6 is the one release where it's needed), and then the main port was built. What am I missing?

Also, what changes haven't been made?

**Comment by mascguy** [comment]
> Looks OK, but I would want confirmation that clang-3.4 still bootstraps OK on 10.6 before merging.

Okay, so would you like me to build clang-3.4 from source, with my updated python27 (and gettext 0.21)?

**Comment by mascguy** [comment]
> Main port shouldn't be rev bumped, and we need verification that this still works for bootstrapping (see previous review.)

Fair enough. I'm in the process of rebuilding clang-3.4 from source, with the updated python27-bootstrap and gettext 0.21, on 10.6.

**Comment by mascguy** [comment]
> > Main port shouldn't be rev bumped, and we need verification that this still works for bootstrapping (see previous review.)
> 
> Fair enough. I'm in the process of rebuilding clang-3.4 from source, with the updated python27-bootstrap and gettext 0.21, on 10.6.

Josh, is that the proper build order, in terms of validating the various changes?

**Comment by mascguy** [comment]
> > Main port shouldn't be rev bumped, and we need verification that this still works for bootstrapping (see previous review.)
> 
> Fair enough. I'm in the process of rebuilding clang-3.4 from source, with the updated python27-bootstrap and gettext 0.21, on 10.6.

clang-3.4 built successfully from source on 10.6, utilizing the updated python27-bootstrap.

Is there any additional validation, etc, that's needed?

**Comment by mascguy** [comment]
Josh, if the clang-3.4 validation build takes care of your concerns, can you please approve this? Or better yet, merge it?

**Comment by mascguy** [comment]
I'm also happy to rebuild additional components from source, if folks would like.

But it would help if folks were more responsive, as this is starting to feel like pulling teeth. And that's not beneficial to our users, nor to MacPorts as whole.

You have a resource at your disposal, who's willing to go the extra mile - and then some - to push this past the finish line. Take advantage of it!

**Comment by jmroot** [comment]
It's been less than a day. Your expectations may need to be adjusted. Most people can't drop everything else they're doing to give you a review immediately, nor should they be expected to.

**Comment by mascguy** [comment]
> It's been less than a day. Your expectations may need to be adjusted. Most people can't drop everything else they're doing to give you a review immediately, nor should they be expected to.

You're right, and perhaps I'm approaching these with a bit too much passion and enthusiasm. Looking forward to seeing these merged though, as the `gettext` upgrade is welcome - and long-overdue.

**Comment by kencu** [comment]
To make sure this works, you need to set up a brand new clean 10.6 macports install, with no ports installed.

It could possibly be a new prefix rather than a new VM or system.

Then enable this PR in your ports repo, and from a completely clean slate, with everything set to build from source and not install any binaries, get to at least clang-3.4 and ideally clang-5.0 by running:
{{{
sudo port -v -s install clang-3.4
}}}

It should run right through and not error or require any manual intervention at any point.

I did this a few months ago, and there was some hiccup so I didn't push this then. But I think it was something else I fixed at that time. And I never came back to it, as this is low on the totem pole for things to take care of on my list at least.

**Comment by mascguy** [comment]
> To make sure this works, you need to set up a brand new clean 10.6 macports install, with no ports installed.
> 
> It could possibly be a new prefix rather than a new VM or system, if `/usr/lib/libc++*` is not installed. Or completely uninstall and remove from your system all the ports you have built, same thing, if you are willing to do that.
> 
> Then enable this PR in your ports repo, and from a completely clean slate, with everything set to build from source and not install any binaries, get to at least clang-3.4 and ideally clang-5.0 by running:
> 
> ```
> sudo port -v -s -N install clang-3.4
> ```
> 
> It should run right through and not error or require any manual intervention at any point.
> 
> I did this a few months ago, and there was some hiccup so I didn't push this then. But I think it was something else I fixed at that time. And I never came back to it, as this is low on the totem pole for things to take care of on my list at least.

Sounds good. I'll clone my existing 10.6 VM, uninstall all ports, and install via your recommendation. Stay Tuned...

**Comment by mascguy** [comment]
One minor concern: It looks like this is going to require a source build of clang-9.0 and friends?

```
$ port rdeps clang-3.4
The following ports are dependencies of clang-3.4 @3.4.2_14:
  cctools
    libunwind-headers
      xz-bootstrap
    llvm-3.4
      libffi
        expect
          automake
            autoconf
              xz
                clang-3.4
                  python27-bootstrap
                    pkgconfig-bootstrap
                      libiconv-bootstrap
                        gperf-bootstrap
                    bzip2
                    expat
                      lzip
                        clang-9.0
                          cmake
                            clang-3.7
                              llvm-3.7
                                libedit
                                  ncurses
                                    libcxx
                                zlib
                                python27
                                  pkgconfig
                                    libiconv
                                      gperf
                                  db48
                                  gettext
                                    libtextstyle
                                    gettext-runtime
                                    gettext-tools-libs
                                  openssl
                                  sqlite3
                                  python_select
                                  python2_select
                                perl5
                                  perl5.28
                                    gdbm
                                      readline
                                llvm_select
                              clang_select
                              ld64
                                ld64-127
                                  libmacho-headers
                            curl
                              libidn2
                                libunistring
                                  texinfo
                                    help2man
                                      perl5.30
                                      p5.30-locale-gettext
                                  libtool
                                    xattr
                                      unzip
                              libpsl
                                python39
                                  python3_select
                                glib2
                                  libxml2
                                    icu
                                  pcre
                              zstd
                                lz4
                              curl-ca-bundle
                            libarchive
                              lzo2
                            libuv
                              legacy-support
                          libomp
                          llvm-9.0
                            xar
                    ncurses-bootstrap
              m4
          tcl
        dejagnu
```

**Comment by mascguy** [comment]
Was everyone aware of the current dependency tree, per my prior comment? I was a bit surprised, and dismayed...

**Comment by mascguy** [comment]
I'd think we'd want to simplify the dependency tree for 10.6, even if that means using an older version of `lzip` [that doesn't require clang-9.0].

**Comment by mascguy** [comment]
Unless I'm missing something - and this certainly looks like a circular dependency issue - I'd think we'd need the following in the `lzip` portfile, similar to what we're doing for `bzip2`:

```
if {${os.platform} eq "darwin" && ${os.major} < 11} {
    # Having the stdlib set to libc++ on 10.6 causes a dependency on a
    # macports-clang compiler to be added, which would be a dep cycle.
    configure.cxx_stdlib
}
```

Yes/No?

**Comment by mascguy** [comment]
That seem fix is needed for the following ports:
- expect
- m4
- tcl

**Comment by mascguy** [comment]
With the four aforementioned ports updated to not depend on `libcxx`, the dependency graph is a lot better.

However, `xz` is still a concern. While there is an `xz-bootstrap` subport, this still looks like a potential circular dependency issue:

```
$ port rdeps clang-3.4
The following ports are dependencies of clang-3.4 @3.4.2_14:
  cctools
    libunwind-headers
      xz-bootstrap
    llvm-3.4
      libffi
        expect
          automake
            autoconf
              xz
                clang-3.4
                  python27-bootstrap
                    pkgconfig-bootstrap
                      libiconv-bootstrap
                        gperf-bootstrap
                    bzip2
                    expat
                      lzip
                    ncurses-bootstrap
                    openssl
                      zlib
                  clang_select
                  ld64
                    ld64-127
                      libmacho-headers
                libiconv
                  gperf
                    libcxx
                gettext
                  libtextstyle
                    ncurses
                  gettext-runtime
                  gettext-tools-libs
              m4
          tcl
        dejagnu
      perl5
        perl5.28
          db48
          gdbm
            readline
      llvm_select
```

**Comment by kencu** [comment]
Well the last time I bootstrapped from bare metal three or four months ago, I did not need to do any such surgery in so many ports... so I would be surprised if things had deteriorated that badly in just a few months.

**Comment by mascguy** [comment]
> Well the last time I bootstrapped from bare metal three or four months ago, I did not need to do any such surgery in so many ports... so I would be surprised if things had deteriorated that badly in just a few months.

Well, boot up a 10.6 VM, and take a look at `port rdeps clang-3.4`. Doesn't look pretty...

**Comment by kencu** [comment]
and we can't really have all those ports installing against libstdc++ on a working 10.6.8 system.

All ports need to link against libc++, except for the bare minimum, usually tucked away in the bootstrap subdirectory, to get bootstrapped to clang-3.4.

clang-3.4 can build at least a basic libc++, and once a basic libc++ is built, everything should built against libc++.

(getting a proper libc++ requires clang-5.0 to be installed, and that unfortunately requires a manual build step as I have not been willing to generate a whole mess more ports that rely on a bootstrap version of libc++)

**Comment by kencu** [comment]
port rdeps is not all that accurate, because all the path: deps mess you up.

it looks like they are needed, but they are not really needed.

just try the build instead, and you will see what I mean -- port rdeps lies to you sometimes.

**Comment by kencu** [comment]
don't forget you have to manually remove ```/usr/lib/libc++*``` from your 10.6.8 system or VM for a proper test.

just uninstalling the libcxx port is not sufficient.

**Comment by mascguy** [comment]
> port rdeps is not all that accurate, because all the path: deps mess you up.
> 
> it looks like they are needed, but they are not really needed.
> 
> just try the build instead, and you will see what I mean -- port rdeps lies to you sometimes.

Ah, that's one of things I didn't consider. Back to building...

**Comment by mascguy** [comment]
Okay, the source build of clang-3.4 and all dependencies completed successfully on 10.6.

We should have at least one other person do the same, before concluding all is well. But looks very encouraging...

**Comment by jmroot** [comment]
> You're right, and perhaps I'm approaching these with a bit too much passion and enthusiasm.

Passion and enthusiasm are great, just be aware that others may have much less free time than you, or it may be the middle of the night for them due to time zone differences, etc. :)

**Comment by jmroot** [comment]
> port rdeps lies to you sometimes.

Or at least, what it says is true "from a certain point of view". As well as the path: deps making things confusing, it considers depends_test, which is not needed just to install a port. That's what's happening with `expect` for one.

> Okay, the source build of clang-3.4 and all dependencies completed successfully on 10.6.

Great.

**Review by jmroot — APPROVED** [review]

**Comment by mascguy** [comment]
Ken, would you be interested in going through a full source build, to validate our results?

**Comment by kencu** [comment]
> Ken, would you be interested in running a full source build of clang-3.4 on 10.6, to validate our results?

Just bogged down right now trying to fix up a few small hiccups that showed up with the new clang-12 build, and then get the build working back to the older systems we need.

**Comment by mascguy** [comment]
> > Ken, would you be interested in running a full source build of clang-3.4 on 10.6, to validate our results?
> 
> Just bogged down right now trying to fix up a few small hiccups that showed up with the new clang-12 build, and then get the build working back to the older systems we need.

No worries, appreciate the help!

**Comment by mascguy** [comment]
Rebased against master

**Comment by pmetzger** [comment]
I believe this is now the oldest open PR. What do we need to do to finish it?

**Comment by mascguy** [comment]
Rebased against master, to pick up latest enhancements from @jmroot.

**Comment by fhgwright** [comment]
Note that a simple `port deps python27-bootstrap` lists `gettext-bootstrap` as a library dependency in the 64-bit 10.6 cases, but not the 32-bit case.

**Comment by mascguy** [comment]
> Note that a simple `port deps python27-bootstrap` lists `gettext-bootstrap` as a library dependency in the 64-bit 10.6 cases, but not the 32-bit case.

Interesting, as that's not what I'm seeing on my 64-bit 10.6 VM:

```
$ port deps python27-bootstrap
Full Name: python27-bootstrap @2.7.18_4
Extract Dependencies: xz-bootstrap
Build Dependencies:   pkgconfig-bootstrap
Library Dependencies: bzip2, expat, libffi, ncurses-bootstrap, openssl, zlib
Runtime Dependencies: 
```

**Comment by fhgwright** [comment]
I can investigate that a little more, but I probably won't have time today.

**Comment by mascguy** [comment]
> I can investigate that a little more, but I probably won't have time today.

No worries Fred, really appreciate you jumping in and helping to validate this!

As for this PR in-general: Assuming that Fred didn't uncover something we overlooked - and we certainly won't commit until we've gotten past that - I'm looking for general input as to whether this PR is ready for merge.

Ideally I'd like to hear from everyone, but it's understandable that folks often have a lot of other things going on. Nonetheless, if the MacPorts Managers - as well as the veterans - can provide input, we'd really appreciate it!

@ryandesign
@jmroot
@raimue
@kencu
@cjones051073

Others are welcome to chime in as well; the more the merrier!

**Comment by fhgwright** [comment]
It turns out that the reason for the difference among the VMs was that I'd missed the tweaks to `macports.conf` and `sources.conf` in two of the three cases.  32 vs. 64 bits was just a coincidence.

But after fixing that and rerunning all three, all three failed, including the 32-bit VM which should  have been unchanged from the previous run.  The error in all three cases is:
`Error: The following dependencies were not installed because all of them have unmet dependencies (likely due to a dependency cycle): python27-bootstrap openssl pkgconfig clang-3.4 clang-3.4 libcxx libiconv gperf`

The duplicate `clang-3.4` is for real.

What did get successfully installed prior to the error was:

`  bzip2 @1.0.8_0 (active)`
`  cctools @949.0.1_1+llvm34 (active)`
`  clang_select @2.2_0 (active)`
`  expat @2.4.1_0 (active)`
`  gperf-bootstrap @3.1_0 (active)`
`  ld64 @3_3+ld64_127 (active)`
`  ld64-127 @127.2_14+llvm34 (active)`
`  libffi @3.4.2_2 (active)`
`  libiconv-bootstrap @1.16_1 (active)`
`  libmacho-headers @949.0.1_0 (active)`
`  libunwind-headers @5.0.1_0 (active)`
`  llvm-3.4 @3.4.2_15 (active)`
`  llvm_select @2_0 (active)`
`  ncurses-bootstrap @6.2_1 (active)`
`  openssl11 @1.1.1l_5 (active)`
`  pkgconfig-bootstrap @0.29.2_0 (active)`
`  xz-bootstrap @5.2.5_1 (active)`
`  zlib @1.2.11_0 (active)`

I have no idea why the 32-bit case worked the first time and fails now, though the failure seems to be consistent across the three VMs.  Repeating the same command in the current state fails almost immediately with the dependency error.

The three (fairly similar) `clang-3.4` logfiles are here:

[MacSL.log](https://github.com/macports/macports-ports/files/7471356/MacSL.log)
[MacSL32.log](https://github.com/macports/macports-ports/files/7471357/MacSL32.log)
[MacSLS.log](https://github.com/macports/macports-ports/files/7471358/MacSLS.log)

**Comment by mascguy** [comment]
> But after fixing that and rerunning all three, all three failed, including the 32-bit VM which should have been unchanged from the previous run. The error in all three cases is: `Error: The following dependencies were not installed because all of them have unmet dependencies (likely due to a dependency cycle): python27-bootstrap openssl pkgconfig clang-3.4 clang-3.4 libcxx libiconv gperf`

I'm not 100% certain, but perhaps could be related to this new ticket, added by @ryandesign 12 hours ago:
https://trac.macports.org/ticket/63784

**Comment by mascguy** [comment]
@fhgwright Fred, now that Josh has fixed the dependency cycle issue, can you check whether everything looks OK now?

**Comment by fhgwright** [comment]
> @fhgwright Fred, now that Josh has fixed the dependency cycle issue, can you check whether everything looks OK now?

Yes, but I won't have time until next week.

**Comment by mascguy** [comment]
Folks, just a quick FYI: A few days ago, I deleted a number of comments on this PR - many of which were my own, but also those of others - which were mostly superfluous. (Things either related to status over time - "can we merge this yet," "rebased against master," etc - as well as a large trail related to [at the time, anyway] my previous misunderstanding of path-style deps, etc.)

However, I also tried to keep the most important ones - such as the earliest ones from Josh and Ryan, related to details regarding what's involved with testing of this, etc. The most recent ones, related to our current state, were also kept.

In any case, this was done with the best of intentions, which is to make it easier for folks to review and follow this discussion in the future.

With all that being said, now that I understand the official policy per @jmroot - which is to hide comments, rather than delete them - I'll take that to heart going forward.

Still learning and adapting, but also trying my best to make a positive difference here. So please try to keep humoring me, while still keeping me on my toes!

Cheers,
-Chris

**Comment by mascguy** [comment]
> Yes, but I won't have time until next week.

Folks, now that the OpenSSL 3 migration is behind us - and the dependency cycle issue has been fixed (thanks @jmroot!) - perhaps we're ready to revisit this.

I'm planning to re-test, starting with a clean MacPorts installation, via my MacBookPro 17 [running 10.6]. With any luck, I'll find time to do that over the next few days.

Fred, do you think you'll be able to find any time over the next week-ish, to re-test as well? (There's certainly no rush, but if you can find some time over the next week or two, it would be great to have another re-validate all of this too!)

I'll report back with my findings, once complete.

**Comment by fhgwright** [comment]
@mascguy 
> Folks, now that the OpenSSL 3 migration is behind us - and the dependency cycle issue has been fixed (thanks @jmroot!) - perhaps we're ready to revisit this.

The OpenSSL 3 fallout is far from over.

> I'm planning to re-test, starting with a clean MacPorts installation, via my MacBookPro 17 [running 10.6]. With any luck, I'll find time to do that over the next few days.
> 
> Fred, do you think you'll be able to find any time over the next week or two, to re-test as well? (There's certainly no rush, but it would be great to have another re-validate all of this too!)

Within that time frame, yes.

**Comment by mascguy** [comment]
> > Fred, do you think you'll be able to find any time over the next week or two, to re-test as well? (There's certainly no rush, but it would be great to have another re-validate all of this too!)
> 
> Within that time frame, yes.

Sounds good, and there's no rush. I've also been preoccupied of late, assisting folks with pre-merge testing/validation of various PRs.

That said, if anyone else is willing to step up and help us validate this, we'd be grateful! @kencu or @ryandesign, would you be interested in assisting with this, over the next month or so? (I'd love to see us merge this prior to January 1st, so that we can finally update `gettext` per the other PR.)

**Comment by fhgwright** [comment]
> Fred, do you think you'll be able to find any time over the next week or two, to re-test as well? (There's certainly no rush, but it would be great to have another re-validate all of this too!)

@mascguy 
I've now retested on the same VMs (10.6 i386, 10.6 x86_64, 10.6 server x86_64), and it works, with no unexpected bootstrap ports (though it still winds up with bootstrap versions of `gperf`, `libiconv`, `ncurses`, `pkgconfig`, and `xz`, not to mention, of course, `python27-bootstrap`).

File: lang/python27/Portfile
Comment (jmroot) [inline lang/python27/Portfile:64]: The revision needs to be 6 now.

**Review by jmroot — CHANGES_REQUESTED** [review]
Should be good to go once the revision is updated.

File: lang/python27/Portfile
Comment (mascguy) [reply lang/python27/Portfile]: Done!

**Comment by mascguy** [comment]
Rebased against master, and updated the revision for the bootstrap subport to six.

Let's wait for the CI jobs to complete though.

**Comment by mascguy** [comment]
All CI jobs have completed successfully. Merging!

**Comment by mascguy** [comment]
As expected, this change built successfully across all of our buildbots. No surprise whatsoever, but great to see regardless!

**Comment by cjones051073** [comment]
@mascguy  @ryandesign Was this PR the cause of a whole bunch of commits being pushed to master (most of which do not follow our format guildelines) ?

**Comment by mascguy** [comment]
> @mascguy @ryandesign Was this PR the cause of a whole bunch of commits being pushed to master (most of which do not follow our format guildelines) ?

Yes. Ultimately I could have gone back and reworded the various commit messages, but was more concerned with merging this 18-month-old PR.

**Comment by mascguy** [comment]
Sorry, didn't realize you commented on this specific PR, which was merged three days ago.

The commits you mentioned were from [PR 7399 - gettext: update to 0.21 with runtime and tools split](https://github.com/macports/macports-ports/pull/7399)

**Comment by cjones051073** [comment]
Well, I think it would then have been a bit better to use the github squash option when merging, to squash all the commits into one well formatted commit message. I don't actually think we needed to see all the various iterations in those commits anyway in the master history.

**Comment by mascguy** [comment]
> Well, I think it would then have been a bit better to use the github squash option when merging, to squash all the commits into one well formatted commit message. I don't actually think we needed to see all the various iterations in those commits anyway in the master history.

Given that there were some key comments from Ryan in the various commits, explaining what was being done and why, I felt it would be better to retain the history.

**Comment by cjones051073** [comment]
Well then, in my opinion some attempt to clean up the commit history should have been made prior to merging. We really didn't need commits like https://github.com/macports/macports-ports/commit/ff70b7f3d4b12977827833e650b2f2773e7ab0ba finding their way into our permanent master history..

**Comment by pmetzger** [comment]
You can keep all the messages when doing a squash, too.

**Comment by mascguy** [comment]
> You can keep all the messages when doing a squash, too.

True. But that can get messy when there are a number of different files, with different applicable comments.

The easiest approach may be to clean things up via an interactive rebase, just before merge.

**Pull request merged**
